README
NAME
    XML::OPML::SimpleGen - create OPML using XML::Simple

VERSION
    version 0.06_01

SYNOPSIS
        require XML::OPML::SimpleGen;

        my $opml = new XML::OPML::SimpleGen();

        $opml->head(
                 title => 'FIFFS Subscriptions',
               );

        $opml->insert_outline(
            group => 'news',  # groups will be auto generated
            text =>  'some feed',
            xmlUrl => 'http://www.somepage.org/feed.xml',
        );

        # insert_outline and add_outline are the same

        $opml->add_group( text => 'myGroup' ); # explicitly create groups
   
        print $opml->to_string;

        $opml->save('somefile.opml');

        $opml->xml_options( $hashref ); # XML::Simple compatible options

        # See XML::OPML's synopsis for more knowledge

DESCRIPTION
    XML::OPML::SimpleGen lets you simply generate OPML documents without
    having too much to worry about. It is a drop-in replacement for
    XML::OPML in regards of generation. As this module uses XML::Simple it
    is rather generous in regards of attribute or element names.

NAME
    XML::OPML::SimpleGen - create OPML using XML::Simple

COMMON METHODS
    new( key => value )
        Creates a new XML::OPML::SimpleGen instance. All key values will be
        used as attributes for the <atom> element. The only thing you might
        want to use here is the version => '1.1', which is default anyway.

    head( key => value )
        XML::OPML compatible head method to change header values.

    id ( )
        Returns (and increments) a counter.

    add_group ( text => 'name' )
        Method to explicitly create a group which can hold multiple outline
        elements.

    insert_outline ( key => value )
        XML::OPML compatible method to add an outline element. See XML::OPML
        for details. The group key is used to put elements in a certain
        group. Non existent groups will be created automagically.

    add_outline ( key => value )
        Alias to insert_outline for XML::OPML compatibility.

    as_string
        Returns the given OPML XML data as a string

    save ( $filename )
        Saves the OPML data to a file

ADVANCED METHODS
    xml_options ( $hashref )
        $hashref may contain any XML::Simple options.

    outline ( $hashref )
        The outline method defines the 'template' for any new outline
        element. You can preset key value pairs here to be used in all
        outline elements that will be generated by XML::OPML::SimpleGen.

    group ( $hashref )
        This method is similar to outline, it defines the template for a
        grouping outline element.

MAINTAINER
    Stephen Cardie "<stephenca@cpan.org>"

REPOSITORY
    <https://github.com/stephenca/XML-OPML-SimpleGen>

CONTRIBUTORS
    KAPPA "<kappa@cpan.org>" contributed a patch to close RT51000
    <https://rt.cpan.org/Public/Bug/Display.html?id=51000>
    gregoa@debian.org contributed a patch to close RT77725
    <https://rt.cpan.org/Public/Bug/Display.html?id=77725>

REPO
      The git repository for this module is at
    L<https://github.com/stephenca/XML-OPML-SimpleGen>

BUGS
    Please report any bugs or feature requests to
    "bug-xml-opml-simlegen@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-OPML-SimleGen>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SEE ALSO
    XML::OPML XML::Simple

AUTHOR
    Marcus Theisen <marcus@thiesen.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Marcus Thiesen.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for XML-OPML-SimpleGen
0.06_01  Tue May 14 22:59:08 2013 +0100
   - Use DateTime instead of POSIX::strfrime() - avoids problems on win32 and
     other platforms which implement strftime differently. (Stephen Cardie)

0.06 
   - Release v0.06. (Stephen Cardie)
   - Fix RT77725 - 03-localefix.t incorrectly fails on first 9 days of month.
(Provided by gregoa [at ] debian.org) (Stephen Cardie)

0.05    Fri May 18 16:04:55 2012 +0100
    - Release v0.05 - includes bugfix for RT51000

0.040_004  Fri May 18 15:26:16 2012 +0100
    - Bug fix - merge KAPPA's fix for RT51000. (Stephen Cardie)

0.040_003  Fri May 18 14:54:13 2012 +0100
    - Developer release - add perlcritic tests to release stage (Stephen
Cardie)

0.040_002  Fri May 18 14:47:59 2012 +0100
    - Developer release  - fix indirect object syntax; was causing perlcritic
failures (Stephen Cardie)

0.040_001  Fri May 18 14:34:40 2012 +0100
    - Developer release - move to Dist::Zilla for managing infrastructure.
(Stephen Cardie)

0.04 2008-02-08 10:27:00 UTC
    - fix improper plan() declaration in t/02-parse.t. (Stephen Cardie)

0.03 2008-02-03 23:25 UTC
    - make 5.6.0 oldest supported Perl version.
    - use version pragma.
    - eliminate dependencies on Date* modules.
    - revise tests to add taint mode, pod coverage and remove un-needed
      includes.

0.02    2005-03-23 17:55 UTC
	- dogh.. had atom still in my head
	. added a test for passing XML::OPMLs parser

0.01    Date/time
        First version, released on an unsuspecting world.



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "create OPML using XML::Simple",
  "AUTHOR" => "Marcus Theisen <marcus\@thiesen.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "XML-OPML-SimpleGen",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "XML::OPML::SimpleGen",
  "PREREQ_PM" => {
    "Class::Accessor" => 0,
    "DateTime" => 0,
    "POSIX" => 0,
    "XML::Simple" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.06_01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





blib/lib/XML/OPML/SimpleGen.pm
package XML::OPML::SimpleGen;

use strict;
use warnings;

use base 'Class::Accessor';

use DateTime;

use POSIX qw(setlocale LC_TIME LC_CTYPE);

__PACKAGE__->mk_accessors(qw|groups xml_options outline group xml_head xml_outlines xml|);

# Version set by dist.ini; do not change here.
our $VERSION = '0.06_01'; # VERSION

sub new {
    my $class = shift;
    my @args = @_;

    my $args = {
	    groups  => {},

	    xml     => {
	        version     => '1.1',
	        @args,
	        },			

	    # XML::Simple options
	    xml_options => {
	        RootName    => 'opml', 
	        XMLDecl     => '<?xml version="1.0" encoding="utf-8" ?>',
	        AttrIndent  => 1,
	    },

	    # default values for nodes
	    outline => {
	        type        => 'rss',
	        version     => 'RSS',
	        text        => '',
	        title       => '',
	        description => '',
	    },

	    group => {
	        isOpen      => 'true',
	    },

	    xml_head        => {},
	    xml_outlines    => [],

	    id              => 1,
    };

    my $self = bless $args, $class;

    # Force locale to 'C' rather than local, then reset after setting times.
    # Fixes RT51000.  Thanks to KAPPA for the patch.
    my $old_loc = POSIX::setlocale(LC_TIME, "C");
    my $ts_ar = [ localtime() ];
    $self->head(
        title => '',
        $self->_date( dateCreated  => $ts_ar ),
        $self->_date( dateModified => $ts_ar ),
    );
    POSIX::setlocale(LC_TIME,$old_loc);

    return $self;
}

sub _date {
  my $self  = shift;
  my $type  = shift; # dateCreated or dateModified.
  my $ts_ar = shift; # e.g [ localtime() ]

  my %arg;
  @arg{qw(second minute hour day month year)} = (
      @{$ts_ar}[0..3],
      $ts_ar->[4]+1,
      $ts_ar->[5]+1900 );
  my $dt = DateTime->new( %arg );
  return ( $type => $dt->strftime('%a, %e %b %Y %H:%M:%S %z') );
}

sub id {
    my $self = shift;
    
    return $self->{id}++;
}

sub head {
    my $self = shift;
    my $data = {@_};

    #this is necessary, otherwise XML::Simple will just generate attributes
    while (my ($key,$value) = each %{ $data }) {
	    $self->xml_head->{$key} = [ $value ];
    }
}

sub add_group {
    my $self = shift;
    my %defaults = %{$self->group};
    my $data = {
        id => $self->id,
        %defaults,
        @_ };
 
    die "Need to define 'text' attribute" unless defined $data->{text};

    $data->{outline} = [];

    push @{$self->xml_outlines}, $data;
    $self->groups->{$data->{text}} = $data->{outline};
}

sub insert_outline {
    my $self = shift;
    my %defaults = %{$self->outline};
    my $data = {
        id => $self->id,
        %defaults,
        @_};

    my $parent = $self->xml_outlines;

    if (exists $data->{group}) {
	    if (exists $self->groups->{$data->{group}}) {
	        $parent = $self->groups->{$data->{group}};
	        delete($data->{group});
	    }
        else {
	        $self->add_group('text' => $data->{group});
	        $self->insert_outline(%$data);
	        return;
	    }
    }

    push @{$parent}, $data;
}

sub add_outline {
    my $self = shift;
    $self->insert_outline(@_);
}

sub as_string {
    my $self = shift;

    require  XML::Simple;
    my $xs = XML::Simple->new();

    return $xs->XMLout( $self->_mk_hashref, %{$self->xml_options} );
}

sub _mk_hashref {
    my $self = shift;

    my $hashref =  {
	    %{$self->xml},
	    head => $self->xml_head,
	    body => { outline => $self->xml_outlines },
    };

    return $hashref;
}

sub save {
    my $self = shift;
    my $filename = shift;

    require  XML::Simple;
    my $xs = XML::Simple->new();

    $xs->XMLout( $self->_mk_hashref, %{$self->xml_options}, OutputFile => $filename );
}

1;

# ABSTRACT:  create OPML using XML::Simple

__END__

=pod

=head1 NAME

XML::OPML::SimpleGen - create OPML using XML::Simple

=head1 VERSION

version 0.06_01

=head1 SYNOPSIS

    require XML::OPML::SimpleGen;

    my $opml = new XML::OPML::SimpleGen();

    $opml->head(
             title => 'FIFFS Subscriptions',
           );

    $opml->insert_outline(
        group => 'news',  # groups will be auto generated
        text =>  'some feed',
        xmlUrl => 'http://www.somepage.org/feed.xml',
    );

    # insert_outline and add_outline are the same

    $opml->add_group( text => 'myGroup' ); # explicitly create groups
   
    print $opml->to_string;

    $opml->save('somefile.opml');

    $opml->xml_options( $hashref ); # XML::Simple compatible options

    # See XML::OPML's synopsis for more knowledge

=head1 DESCRIPTION

XML::OPML::SimpleGen lets you simply generate OPML documents
without having too much to worry about. 
It is a drop-in replacement for XML::OPML
in regards of generation. 
As this module uses XML::Simple it is rather
generous in regards of attribute or element names.

=head1 NAME

XML::OPML::SimpleGen - create OPML using XML::Simple

=head1 COMMON METHODS

=over

=item new( key => value )

Creates a new XML::OPML::SimpleGen instance. All key values will be
used as attributes for the <atom> element. The only thing you might
want to use here is the version => '1.1', which is default anyway.

=item head( key => value ) 

XML::OPML compatible head method to change header values. 

=item id ( )

Returns (and increments) a counter.

=item add_group ( text => 'name' )

Method to explicitly create a group which can hold multiple outline
elements.

=item insert_outline ( key => value )

XML::OPML compatible method to add an outline element. See
L<XML::OPML> for details. The group key is used to put elements in a
certain group. Non existent groups will be created automagically. 

=item add_outline ( key => value )

Alias to insert_outline for XML::OPML compatibility.

=item as_string 

Returns the given OPML XML data as a string

=item save ( $filename )

Saves the OPML data to a file

=back

=head1 ADVANCED METHODS

=over

=item xml_options ( $hashref ) 

$hashref may contain any XML::Simple options.

=item outline ( $hashref )

The outline method defines the 'template' for any new outline
element. You can preset key value pairs here to be used
in all outline elements that will be generated by XML::OPML::SimpleGen.

=item group ( $hashref )

This method is similar to outline, it defines the template for a
grouping outline element. 

=back

=head1 MAINTAINER 

Stephen Cardie C<< <stephenca@cpan.org> >>

=head1 REPOSITORY

L<https://github.com/stephenca/XML-OPML-SimpleGen>

=head1 CONTRIBUTORS

=over 4

=item KAPPA C<< <kappa@cpan.org> >> contributed a patch to close RT51000
L<https://rt.cpan.org/Public/Bug/Display.html?id=51000>

=item gregoa@debian.org contributed a patch to close RT77725
L<https://rt.cpan.org/Public/Bug/Display.html?id=77725>

=back

=head1 REPO

  The git repository for this module is at
L<https://github.com/stephenca/XML-OPML-SimpleGen>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-xml-opml-simlegen@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=XML-OPML-SimleGen>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SEE ALSO

L<XML::OPML> L<XML::Simple>

=head1 AUTHOR

Marcus Theisen <marcus@thiesen.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Marcus Thiesen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
